Add Clipping to displaybuffer

nielsnl68 opened the pull request:
# What does this implement/fix?

This per extends the displaybuffer with the option to clip parts of the screen where the changes are allowed.


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
display:
  - platform: ili9341
    model: TFT 2.4
    cs_pin: 14
    dc_pin: 27
    led_pin: 32  
    reset_pin: 33

    lambda: |-
      it.print(20, 20, id(my_red), id(my_red), TextAlign::TOP_LEFT, "Hello World!");
      it.push_clipping(30, 25, 100, 40);
      it.print(20, 20, id(my_red), id(my_blue), TextAlign::TOP_LEFT, "Hello World!");
      it.pop_clipping();

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nielsnl68 committed:
adding Clipping support to the displaybuffer
- add rect structure

nielsnl68 committed:
Merge branch 'nvds-displaybuffer-clipping' of https://github.com/nielsnl68/esphome into nvds-displaybuffer-clipping

nielsnl68 committed:
removed unused define

nielsnl68 committed:
add missing property for storing the clipped areas

nielsnl68 committed:
Merge branch 'dev' into nvds-displaybuffer-clipping

nielsnl68 committed:
include log header

jesserockz commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -392,6 +395,43 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
 #endif
 
+void DisplayBuffer::push_clipping(Rect rect) {
+  // ESP_LOGW(TAG, "set: Push new clipping");
```
```suggestion
```

nielsnl68 replied:
I have moved the function code to the cpp file now.

jesserockz commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -392,6 +395,43 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
+  if (this->clipping_rectangle_.empty()) {
+    ESP_LOGE(TAG, "add: Clipping is not set.");
+  } else {
+    // ESP_LOGW(TAG, "add: join new clipping");
```
```suggestion
```

jesserockz commented on esphome/components/display/display_buffer.h:
```diff
@@ -374,6 +468,72 @@ class DisplayBuffer {
+  ///                          of contract the height (if negative)
+  ///
+  /// \return new rect with resized dimensions
+  ///
```
???

nielsnl68 replied:
thanks, i have removed it

jesserockz commented on esphome/components/display/display_buffer.h:
```diff
@@ -100,6 +101,98 @@ enum DisplayRotation {
+  //  this        x--------------w
+  inline void info(const std::string &prefix = "rect info:") {
+    if (this->is_set()) {
+      ESP_LOGI("Rect", "%s [%3d,%3d,%3d,%3d]", prefix.c_str(), this->x, this->y, this->w, this->h);
```
Log calls need to be in the cpp file

nielsnl68 replied:
I moved the functions now to the cpp file

jesserockz commented on esphome/components/display/display_buffer.h:
```diff
@@ -4,7 +4,8 @@
 #include "display_color_utils.h"
-
+#include "esphome/core/helpers.h"
+#include "esphome/core/log.h"
```
This needs to go in the cpp file

nielsnl68 replied:
done

nielsnl68 committed:
Move Rect method's code to cpp file
- removed obsolete remarks

nielsnl68 committed:
fixed reported issues

nielsnl68 committed:
make Rect class methods public

nielsnl68 committed:
clang fix

nielsnl68 committed:
Merge branch 'dev' into nvds-displaybuffer-clipping

jesserockz committed:
Remove commented code

nielsnl68 committed:
Renaming clipping methods

nielsnl68 committed:
Multiple changes:
- replaced 32766 with VALUE_NO_SET
- fixed the way *_clipping(left, top, right, bottom) is stored
- add `is_clipping();`
- make sure that all clipped region are closed after `do_update_()`
- rename de parameters for `Rect::expand();`

nielsnl68 committed:
remove unneeded space

nielsnl68 committed:
replace define with static const uint8_t

nielsnl68 committed:
correcting my copy paste mistake

nielsnl68 committed:
Merge branch 'dev' into nvds-displaybuffer-clipping
